apps/meteor/app/message-pin/server/pinMessage.ts

```diff
@@ -43,108 +43,101 @@ declare module '@rocket.chat/ddp-client' {
 	}
 }
 
-Meteor.methods<ServerMethods>({
-	async pinMessage(message, pinnedAt) {
-		check(message._id, String);
-
-		const userId = Meteor.userId();
-		if (!userId) {
-			throw new Meteor.Error('error-invalid-user', 'Invalid user', {
-				method: 'pinMessage',
-			});
-		}
+export async function pinMessage(originalMessage: IMessage, userId: string, pinnedAt?: Date) {
+	if (!settings.get('Message_AllowPinning')) {
+		throw new Meteor.Error('error-action-not-allowed', 'Message pinning not allowed', {
+			method: 'pinMessage',
+			action: 'Message_pinning',
+		});
+	}
 
-		if (!settings.get('Message_AllowPinning')) {
-			throw new Meteor.Error('error-action-not-allowed', 'Message pinning not allowed', {
-				method: 'pinMessage',
-				action: 'Message_pinning',
-			});
-		}
+	if (originalMessage.pinned) {
+		return originalMessage;
+	}
 
-		let originalMessage = await Messages.findOneById(message._id);
-		if (originalMessage == null || originalMessage._id == null) {
-			throw new Meteor.Error('error-invalid-message', 'Message you are pinning was not found', {
-				method: 'pinMessage',
-				action: 'Message_pinning',
-			});
-		}
+	if (!(await hasPermissionAsync(userId, 'pin-message', originalMessage.rid))) {
```

Comment:
Hey, I think we should have the check for already pinned message after this part. 
If the user does not have permission and the message is already pinned, they will get the original message which they should not.